src/server/mod.rs

```diff
@@ -336,6 +405,381 @@ impl<T, B> Service for HttpService<T>
     }
 }
 
+struct UpgradableHttp<P, B = ::Chunk> {
+    inner: Http<B>,
+    _marker: PhantomData<P>,
+}
+
+type UpgradablePayload<T, B, P> = Result<(__ProtoTransport<T, B>, P), ::Error>;
+type UpgradableReceiver<T, B, P> = Receiver<UpgradablePayload<T, B, P>>;
+type UpgradableSender<T, B, P> = Sender<UpgradablePayload<T, B, P>>;
+
+type UpgradableResponseHead<P> = Result<__ProtoResponse, (P, Option<__ProtoResponse>)>;
+
+impl<T, B, P> ServerProto<(T, UpgradableSender<T, B, P>)> for UpgradableHttp<P, B>
```

Comment:
That all of these boilerplate impls are needed is a bummer. Sorry to have to write that XD